source/javascripts/core/services/StackAndMachineService.spec.ts

```diff
@@ -413,33 +489,91 @@ describe('changeStackAndMachine', () => {
         machineTypeId: 'standard',
         availableStacks: stacks,
         availableMachineTypes: machines,
-        defaultStackId: 'osx-xcode-16',
+        projectStackId: 'osx-xcode-16',
       });
 
       expect(result).toEqual({ stackId: 'osx-xcode-16', machineTypeId: '' });
     });
   });
 
   describe('stack OS changes', () => {
-    it('keeps a specific stack with the default machine type', () => {
+    it('keeps a specific stack with empty machine type', () => {
       const result = StackAndMachineService.changeStackAndMachine({
         stackId: 'linux-ubuntu-22',
         machineTypeId: '',
         availableStacks: stacks,
         availableMachineTypes: machines,
-        defaultStackId: 'osx-xcode-16',
+        projectStackId: 'osx-xcode-16',
       });
 
       expect(result).toEqual({ stackId: 'linux-ubuntu-22', machineTypeId: '' });
     });
 
+    it('keeps a specific stack with project machine type', () => {
+      const result = StackAndMachineService.changeStackAndMachine({
+        stackId: 'osx-xcode-15',
+        machineTypeId: '',
+        availableStacks: stacks,
+        availableMachineTypes: machines,
+        projectStackId: 'osx-xcode-16',
+        machineFallbackOptions: {
+          defaultMachineTypeIdOfOSs: {
+            linux: 'elite',
+            osx: 'mac-m2',
+          },
+          projectMachineTypeId: 'mac-m1',
+        },
+      });
+
+      expect(result).toEqual({ stackId: 'osx-xcode-15', machineTypeId: 'mac-m1' });
+    });
+
+    it('keeps a specific stack with default machine type of OS', () => {
+      const result = StackAndMachineService.changeStackAndMachine({
+        stackId: 'linux-ubuntu-22',
+        machineTypeId: '',
+        availableStacks: stacks,
+        availableMachineTypes: machines,
+        projectStackId: 'osx-xcode-16',
+        machineFallbackOptions: {
+          defaultMachineTypeIdOfOSs: {
+            linux: 'elite',
+            osx: 'mac-m1',
+          },
+          projectMachineTypeId: 'mac-m1',
+        },
+      });
+
+      expect(result).toEqual({ stackId: 'linux-ubuntu-22', machineTypeId: 'elite' });
+    });
+
+    it('keeps a specific stack with the first available machine type', () => {
+      const result = StackAndMachineService.changeStackAndMachine({
+        stackId: 'linux-ubuntu-22',
+        machineTypeId: '',
+        availableStacks: stacks,
+        availableMachineTypes: machines,
+        projectStackId: 'osx-xcode-16',
+        machineFallbackOptions: {
+          defaultMachineTypeIdOfOSs: {
+            linux: 'large',
+            osx: 'mac-m2',
+          },
+          projectMachineTypeId: 'mac-m1',
+        },
+      });
+
+      expect(result).toEqual({ stackId: 'linux-ubuntu-22', machineTypeId: 'standard' });
+    });
```

Comment:
Let's move these into a new describe block, because they all test the machineFallbackOptions?